### PR TITLE
Compare performance of  tt-lang vs ttml kernels

### PR DESCRIPTION
**Pull request by lgalasTT**
This PR is intended to discuss the performance comparison results of the TT-Lang and TTML kernels.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:lgalasTT/benchmark_ttlang_vs_ttml_kernels)

### CODE REVIEW

File: tt-train/sources/ttml/autograd/tensor.cpp [hunk 1]
```diff
@@ -164,4 +164,8 @@ uint32_t Tensor::get_rank() const {
     return get_shape().rank();
 }
 
+void Tensor::deallocate_storage() {
+    m_value.deallocate_storage();
+}
```
Comment (philei-tt) [inline tt-train/sources/ttml/autograd/tensor.cpp]: Why?? What's wrong with RAII? If you really need to deallocate it asap, you could just do tensor->value().deallocate()

File: tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_device_operation_types.hpp [hunk 2]
```diff
@@ -11,6 +11,8 @@ namespace ttml::metal::ops::rmsnorm_bw::device {
 // Attributes for the backward operation (add more if needed)
 struct operation_attributes_t {
     float epsilon = 1e-6F;
+    // Maximum worker cores to use (clamped to the device compute grid). 0 means use the full grid.
+    uint32_t max_num_cores = 0U;
 };
```
Comment (philei-tt) [inline tt-train/sources/ttml/metal/ops/rmsnorm_bw/device/rmsnorm_bw_device_operation_types.hpp:14]: Why would you want to use less cores?

File: tt-train/sources/ttml/nanobind/nb_ops.cpp [hunk 3]
```diff
@@ -366,6 +367,36 @@ void py_module(nb::module_& m) {
             nb::arg("tensor"),
             nb::arg("gamma"),
             nb::arg("epsilon"));
+
+        // Low-level RMSNorm backward (matches tt-train C++ `ttml::metal::rmsnorm_bw`), for microbenchmarks.
+        py_rmsnorm.def(
+            "rmsnorm_bw",
```
Comment (philei-tt) [inline tt-train/sources/ttml/nanobind/nb_ops.cpp:408]: Are you going to remove this before merging? If no, I suggest putting this under some "detail" submodule or something like that

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+TTLANG:
+    256x384_KernelBw_TTL          Time_us=    249.63  GB/s=   2.371  Elems_M=0.0983
+    256x2048_KernelBw_TTL         Time_us=    250.13  GB/s=  12.611  Elems_M=0.5243
+    1024x2048_KernelBw_TTL        Time_us=    393.97  GB/s=  31.965  Elems_M=2.0972
+    2048x2048_KernelBw_TTL        Time_us=    506.14  GB/s=  49.745  Elems_M=4.1943
+    2048x5632_KernelBw_TTL        Time_us=    817.26  GB/s=  84.713  Elems_M=11.5343
+    4096x4096_KernelBw_TTL        Time_us=   1033.92  GB/s=  97.385  Elems_M=16.7772
+    4096x11008_KernelBw_TTL       Time_us=   2338.71  GB/s= 115.698  Elems_M=45.0888
+    8192x8192_KernelBw_TTL        Time_us=   3380.75  GB/s= 119.116  Elems_M=67.1089
+
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py:33]: It's very weird to put benchmark results in docstring, no? Just make markdown if you want to somehow track this, but it's better to just leave them in pr description imo

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+KERNEL_BW_TTML_MAX_CORES = 0
+
+SHAPES = (
+    # ([1, 1, 256, 384], "256x384"),
+    # ([1, 1, 256, 2048], "256x2048"),
+    # ([1, 1, 256, 11008], "1024x2048"),
+    # ([2, 1, 1024, 2048], "2048x2048"),
+    # ([2, 1, 1024, 5632], "2048x5632"),
+    # ([4, 1, 1024, 4096], "4096x4096"),
+    # ([4, 1, 1024, 11008], "4096x11008"),
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py]: Commented code

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+
+    return ttl_rmsnorm_bw
+
+
+from ttml.autograd import AutoContext, Tensor, create_tensor
+from ttml import ops
+
+
+def _open_mesh_for_kernel_bw(ctx: AutoContext, bw_kernel: str) -> None:
+    """Open mesh; TTL paths reserve L1 so the kernel-config ringbuffer fits large TTL programs."""
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py]: How much L1 does it need? I don't quite get, they reserve some L1 buffer throughout the lifetime of a whole metal runtime to fit bigger program or what?

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+            tag = "TTL (ttl_rmsnorm_bw.py, 1c)"
+        print(f"Mode: KernelBw / {tag} (backward-only, RMS from torch forward on host)\n")
+        if bw_kernel == "ttl":
+            print(
+                "[kernel_bw] ttl: ttl.operation(grid=(1, 1)) — 1 logical grid core per device launch.\n",
+                flush=True,
+            )
+
+        for shape, name in SHAPES:
+            seed = _seed_for_name(name)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py:235]: Why do you need different seed for different shapes?

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+            "ttml.ops.rmsnorm has neither rmsnorm_bw nor metal_bw. Rebuild _ttml (tt-train) after nb_ops.cpp changes."
+        )
+
+    def call(x, g, r, d):
+        return fn(x, g, r, d, KERNEL_BW_TTML_MAX_CORES)
+
+    return call
+
+
+def _jit_disk_help_if_enospc(exc: BaseException) -> None:
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py]: waaay overkill tbh

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+                flush=True,
+            )
+
+        for shape, name in SHAPES:
+            seed = _seed_for_name(name)
+            rng = np.random.default_rng(seed)
+            b, _, s_len, c = shape
+            x_np = rng.uniform(-1.0, 1.0, (b, 1, s_len, c)).astype(np.float32)
+            g_np = rng.uniform(-1.0, 1.0, (1, 1, 1, c)).astype(np.float32)
+            d_np = rng.uniform(-1.0, 1.0, (b, 1, s_len, c)).astype(np.float32)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py:240]: (nit) maybe ttml.ops.rand could be used

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+                out_da = ttl_mod._to_dev(torch.zeros(rows_p, cols_p, dtype=torch.bfloat16), mesh)
+                out_dg = ttl_mod._to_dev(torch.zeros(rows_p, cols_p, dtype=torch.bfloat16), mesh)
+                ttl_device_tensors.extend((x_p, g_p, rms_p, dL_p, out_da, out_dg))
+
+                def fn(x, g, r, d):
+                    return k(x, g, r, d, out_da, out_dg)
+
+                def run_step() -> None:
+                    fn(x_p, g_p, rms_p, dL_p)
+                    ttnn.synchronize_device(mesh)
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py:266]: syncronization in the loop means you also include dispatch and sync overhead in your measurements, which is usually hidden in real workflow. Those might be not negligible, especially for small shapes . I suggest rewriting this logic to sync once at the end of NUM_MEASURE loop.

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+    except TypeError:
+        ctx.open_device((1, 1))
+
+
+RMSNORM_EPS = 0.0078125
+NUM_WARMUP = 5
+NUM_MEASURE = 50
+
+# Match ``ttl_rmsnorm_bw`` ``@ttl.operation(grid=(1, 1))`` when benchmarking vs ttml metal path.
+KERNEL_BW_TTML_MAX_CORES = 0
```
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py]: I don't get why you added a parameter to select number of cores, only to hardcode it to 0 meaning all cores. Wasn't this a default behaviour in rmsnorm bw kernel?

File: tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py
Comment (philei-tt) [inline tt-train/sources/ttml/ttlang/rmsnorm_bw_benchmark.py:1]: I would love to see tracy measurements. Wall-time benchmarks like that are sometimes inaccurate, so it's good to make sure that tracy measurements are similar

File: tt-train/tests/benchmark/rmnsnorm_bw_benchmark.cpp
Comment (philei-tt) [inline tt-train/tests/benchmark/rmnsnorm_bw_benchmark.cpp:1]: Why do you need a second cpp benchmark? Also it measures only metal rmsnorm, right?
